commit 1a9bdcf5c1af1c3567d803dbc206bae65d8cd875
Author: Tim Theisen <tim@cs.wisc.edu>
Date:   Mon Apr 26 16:31:13 2021 -0500

    Squashed commit of the following:
    
    commit bbbe6e5fd942cd1421ef4456074c420f3c0a08de
    Author: Jason Patton <jpatton@cs.wisc.edu>
    Date:   Mon Apr 26 16:02:54 2021 -0500
    
        Promoted aud config to mandatory in oauth config HTCONDOR-445
    
    commit 003a9e1e5f2c2f122015d2185c725921c400f436
    Author: Derek Weitzel <dweitzel@cse.unl.edu>
    Date:   Mon Apr 26 14:22:13 2021 -0500
    
        Update src/condor_credd/condor_credmon_oauth/credmon/CredentialMonitors/LocalCredmon.py
    
        Co-authored-by: jasoncpatton <jason.c.patton@gmail.com>
    
    commit 0731b6caa04093a221c7a7fd32d429d79590b9b7
    Author: Derek Weitzel <djw8605@gmail.com>
    Date:   Mon Apr 26 13:54:05 2021 -0500
    
        Add audience and version as configuration options for the local credmon

diff --git a/src/condor_credd/condor_credmon_oauth/credmon/CredentialMonitors/LocalCredmon.py b/src/condor_credd/condor_credmon_oauth/credmon/CredentialMonitors/LocalCredmon.py
index c256a86..c9d0a2d 100644
--- a/src/condor_credd/condor_credmon_oauth/credmon/CredentialMonitors/LocalCredmon.py
+++ b/src/condor_credd/condor_credmon_oauth/credmon/CredentialMonitors/LocalCredmon.py
@@ -26,6 +26,8 @@ class LocalCredmon(OAuthCredmon):
         self.authz_template = "read:/user/{username} write:/user/{username}"
         self.token_lifetime = 60*20
         self.token_use_json = True
+        self.token_aud = ""
+        self.token_ver = "scitokens:2.0"
         if htcondor != None:
             self._private_key_location = htcondor.param.get('LOCAL_CREDMON_PRIVATE_KEY', "/etc/condor/scitokens-private.pem")
             if self._private_key_location != None and os.path.exists(self._private_key_location):
@@ -39,6 +41,8 @@ class LocalCredmon(OAuthCredmon):
             self.authz_template = htcondor.param.get("LOCAL_CREDMON_AUTHZ_TEMPLATE", self.authz_template)
             self.token_lifetime = htcondor.param.get("LOCAL_CREDMON_TOKEN_LIFETIME", self.token_lifetime)
             self.token_use_json = htcondor.param.get("LOCAL_CREDMON_TOKEN_USE_JSON", self.token_use_json)
+            self.token_aud = htcondor.param.get("LOCAL_CREDMON_TOKEN_AUDIENCE", self.token_aud)
+            self.token_ver = htcondor.param.get("LOCAL_CREDMON_TOKEN_VERSION", self.token_ver)
         else:
             self._private_key_location = None
         if not self.token_issuer:
@@ -58,7 +62,16 @@ class LocalCredmon(OAuthCredmon):
         token.update_claims({'sub': username})
         user_authz = self.authz_template.format(username=username)
         token.update_claims({'scope': user_authz})
-        token.update_claims({'ver': 'scitokens:2.0'})
+
+        # Only set the version if we have one.  No version is valid, and implies scitokens:1.0
+        if self.token_ver:
+            token.update_claims({'ver': self.token_ver})
+
+        # Convert the space separated list of audiences to a proper list
+        # No aud is valid for scitokens:1.0 tokens.  Also, no resonable default.
+        aud_list = self.token_aud.strip().split()
+        if aud_list:
+            token.update_claims({'aud': aud_list})
 
         # Serialize the token and write it to a file
         try:
diff --git a/src/condor_credd/condor_credmon_oauth/examples/config/condor/40-oauth-credmon.conf b/src/condor_credd/condor_credmon_oauth/examples/config/condor/40-oauth-credmon.conf
index 7f2b568..b4e5638 100644
--- a/src/condor_credd/condor_credmon_oauth/examples/config/condor/40-oauth-credmon.conf
+++ b/src/condor_credd/condor_credmon_oauth/examples/config/condor/40-oauth-credmon.conf
@@ -1,7 +1,6 @@
 
 #####################################################################################
-# MANDATORY: To setup the credmon, you must uncomment these lines:
-#
+# MANDATORY: To setup the credmon, you must uncomment this line:
 use feature : OAUTH
 
 # MANDATORY for local credmon: Uncomment these lines to have submitters always
@@ -15,6 +14,14 @@ use feature : OAUTH
 # LOCAL_CREDMON_PROVIDER_NAME = scitokens
 # SEC_PROCESS_SUBMIT_TOKENS = false
 
+# MANDATORY for local credmon if using scitokens:2.0 tokens (the default version
+# as of HTCondor 8.9.13, see LOCAL_CREDMON_TOKEN_VERSION below)
+# Audience (aud) claim that should be set in a locally issued token.
+# The audience is a shared value between the token issuer and the service
+# verifying the token. It is a space-separated list, will be converted to a JSON
+# list object in the token.
+# LOCAL_CREDMON_TOKEN_AUDIENCE = https://example.com https://anotherserver.edu
+
 # MANDATORY for enabling the transfer of credentials from submit host
 # to execute hosts, if encryption is not already enabled.
 # SEC_DEFAULT_ENCRYPTION = REQUIRED
@@ -57,6 +64,10 @@ use feature : OAUTH
 # or as bare strings (LOCAL_CREDMON_TOKEN_USE_JSON = false)
 # LOCAL_CREDMON_TOKEN_USE_JSON = false
 
+# Version (ver) claim that should be set in a locally issued token
+# Default: scitokens:2.0
+# LOCAL_CREDMON_TOKEN_VERSION = scitokens:2.0
+
 # Override the location of the credential directory, credmon daemon, or credmon log
 # SEC_CREDENTIAL_DIRECTORY_OAUTH = /var/lib/condor/oauth_credentials
 # CREDMON_OAUTH_LOG = $(LOG)/CredMonOAuthLog
diff --git a/src/condor_utils/param_info.in b/src/condor_utils/param_info.in
index 1cd2caf..69be543 100644
--- a/src/condor_utils/param_info.in
+++ b/src/condor_utils/param_info.in
@@ -5526,6 +5526,15 @@ type=path
 default=true
 type=bool
 
+[LOCAL_CREDMON_TOKEN_AUDIENCE]
+default=
+type=string
+
+[LOCAL_CREDMON_TOKEN_VERSION]
+default=scitokens:2.0
+type=string
+
+
 # standard API endpoints for common OAuth token providers
 [BOX_AUTHORIZATION_URL]
 default=https://account.box.com/api/oauth2/authorize
